File: cameo-api/src/main/java/fr/ill/ics/cameo/impl/ServicesImpl.java
/*
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under the EUPL, Version 1.1 only (the "License");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * http://joinup.ec.europa.eu/software/page/eupl
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and
 * limitations under the Licence.
 */

package fr.ill.ics.cameo.impl;

import org.json.simple.JSONObject;
import org.json.simple.parser.ParseException;

import fr.ill.ics.cameo.ConnectionTimeout;
import fr.ill.ics.cameo.EventStreamSocket;
import fr.ill.ics.cameo.SocketException;
import fr.ill.ics.cameo.UnexpectedException;
import fr.ill.ics.cameo.Zmq;
import fr.ill.ics.cameo.Zmq.Socket;
import fr.ill.ics.cameo.messages.JSON;
import fr.ill.ics.cameo.messages.Message;
import fr.ill.ics.cameo.strings.Endpoint;

public class ServicesImpl {

	protected Endpoint serverEndpoint;
	protected int[] serverVersion = new int[3];
	protected int statusPort;
	protected Zmq.Context context;
	protected int timeout = 0; // default value because of ZeroMQ design
	protected RequestSocket requestSocket;
	protected JSON.ConcurrentParser parser = new JSON.ConcurrentParser();
	
	/**
	 * Initializes the context and the request socket. The serverEndpoint must have been set.
	 */
	final protected void init() {
		this.context = new Zmq.Context();
		this.requestSocket = this.createRequestSocket(serverEndpoint.toString());
	}
	
	public int getTimeout() {
		return timeout;
	}

	public void setTimeout(int timeout) {
		this.timeout = timeout;
	}
	
	public Endpoint getEndpoint() {
		return serverEndpoint;
	}
		
	public int[] getVersion() {
		return serverVersion;
	}
	
	public Endpoint getStatusEndpoint() {
		return serverEndpoint.withPort(statusPort);
	}
	
	public void destroySocket(Socket socket) {
		context.destroySocket(socket);
	}
	
	public void terminate() {
		
		// Terminate the request socket.
		requestSocket.terminate();
		
		// destroying the context
		context.destroy();
	}
	
	public JSONObject parse(Zmq.Msg reply) throws ParseException {
		return parser.parse(Message.parseString(reply.getFirstData()));
	}
	
	public JSONObject parse(byte[] data) throws ParseException {
		return parser.parse(Message.parseString(data));
	}
	
	/**
	 * test connection with server
	 * @param timeout
	 * 
	 */
	public boolean isAvailable(int overrideTimeout) {

		Zmq.Msg request = createSyncRequest();
		Zmq.Msg reply = null;
		try {
			reply = requestSocket.request(request, overrideTimeout);
			reply.destroy();
			request.destroy();
			return true;

		} catch (ConnectionTimeout e) {
			// do nothing, timeout
		} catch (Exception e) {
			// do nothing
		}
		
		return false;
	}
	
	protected void sendSync() {
		
		Zmq.Msg request = createSyncRequest();
		try {
			Zmq.Msg reply = requestSocket.request(request);
			reply.destroy();
			request.destroy();

		} catch (ConnectionTimeout e) {
			// do nothing
		}
	}
	
	protected void sendSyncStream(String name) {
		
		Zmq.Msg request = createSyncStreamRequest(name);
		try {
			Zmq.Msg reply = requestSocket.request(request);
			reply.destroy();
			request.destroy();

		} catch (ConnectionTimeout e) {
			// do nothing
		}
	}

	protected void retrieveServerVersion() {
		
		Zmq.Msg request = createVersionRequest();
		Zmq.Msg reply = requestSocket.request(request);
		
		try {
			// Get the JSON response object.
			JSONObject response = parse(reply);
		
			serverVersion[0] = JSON.getInt(response, Message.VersionResponse.MAJOR);
			serverVersion[1] = JSON.getInt(response, Message.VersionResponse.MINOR);
			serverVersion[2] = JSON.getInt(response, Message.VersionResponse.REVISION);
		}
		catch (ParseException e) {
			throw new UnexpectedException("Cannot parse response");
		}
	}
	
	/**
	 * 
	 * @throws ConnectionTimeout 
	 */
	protected EventStreamSocket openEventStream() {

		Zmq.Msg request = createStreamStatusRequest();
		
		Zmq.Msg reply = requestSocket.request(request);
		JSONObject response;
		
		try {
			// Get the JSON response object.
			response = parse(reply);
		}
		catch (ParseException e) {
			throw new UnexpectedException("Cannot parse response");
		}
		
		// Prepare our subscriber.
		Zmq.Socket subscriber = context.createSocket(Zmq.SUB);
		
		statusPort = JSON.getInt(response, Message.RequestResponse.VALUE);
		
		subscriber.connect(getStatusEndpoint().toString());
		subscriber.subscribe(Message.Event.STATUS);
		subscriber.subscribe(Message.Event.RESULT);
		subscriber.subscribe(Message.Event.PUBLISHER);
		subscriber.subscribe(Message.Event.PORT);
		subscriber.subscribe(Message.Event.KEYVALUE);
		
		String cancelEndpoint = "inproc://cancel." + CancelIdGenerator.newId();
		
		subscriber.connect(cancelEndpoint);
		subscriber.subscribe(Message.Event.CANCEL);
		
		// polling to wait for connection
		Zmq.Poller poller = context.createPoller(subscriber);
		
		while (true) {
			
			// the server returns a STATUS message that is used to synchronize the subscriber
			sendSync();

			// return at the first response.
			if (poller.poll(100)) {
				break;
			}
		}
		
		Zmq.Socket cancelPublisher = context.createSocket(Zmq.PUB);
		cancelPublisher.bind(cancelEndpoint);
		
		return new EventStreamSocket(this, subscriber, cancelPublisher);
	}
	
	protected RequestSocket createRequestSocket(String endpoint) throws SocketException {
		
		RequestSocket requestSocket = new RequestSocket(context, timeout);
		requestSocket.connect(endpoint);
		
		return requestSocket;
	}
	
	protected Zmq.Msg message(JSONObject object) {
		
		Zmq.Msg message = new Zmq.Msg();
		message.add(Message.serialize(object));

		return message;
	}
	
	/**
	 * create init request
	 * 
	 * @return
	 */
	protected Zmq.Msg createSyncRequest() {
		
		JSONObject request = new JSONObject();
		request.put(Message.TYPE, Message.SYNC);
		
		return message(request);
	}
	

	protected Zmq.Msg createSyncStreamRequest(String name) {
		
		JSONObject request = new JSONObject();
		request.put(Message.TYPE, Message.SYNC_STREAM);
		request.put(Message.SyncStreamRequest.NAME, name);
		
		return message(request);
	}
	
	/**
	 * create version request
	 * 
	 * @return request
	 */
	protected Zmq.Msg createVersionRequest() {

		JSONObject request = new JSONObject();
		request.put(Message.TYPE, Message.IMPL_VERSION);

		return message(request);
	}
	
	/**
	 * create show status request
	 * 
	 * @return request
	 */
	protected Zmq.Msg createStreamStatusRequest() {
		
		JSONObject request = new JSONObject();
		request.put(Message.TYPE, Message.STATUS);

		return message(request);
	}
	
	/**
	 * create getStatus request
	 * 
	 * @param text
	 * @return
	 */
	protected Zmq.Msg createGetStatusRequest(int id) {
		
		JSONObject request = new JSONObject();
		request.put(Message.TYPE, Message.GET_STATUS);
		request.put(Message.GetStatusRequest.ID, id);

		return message(request);	
	}
	
	/**
	 * create startedUnmanaged request
	 * @param pid 
	 * 
	 * @param text
	 * @return
	 */
	protected Zmq.Msg createAttachUnmanagedRequest(String name, long pid) {
		
		JSONObject request = new JSONObject();
		request.put(Message.TYPE, Message.ATTACH_UNMANAGED);
		request.put(Message.AttachUnmanagedRequest.NAME, name);
		request.put(Message.AttachUnmanagedRequest.PID, pid);
		
		return message(request);
	}
	
	/**
	 * create terminatedUnmanaged request
	 * 
	 * @param text
	 * @return
	 */
	protected Zmq.Msg createDetachUnmanagedRequest(int id) {
		
		JSONObject request = new JSONObject();
		request.put(Message.TYPE, Message.DETACH_UNMANAGED);
		request.put(Message.DetachUnmanagedRequest.ID, id);
		
		return message(request);
	}
	
}

File: cameo-api-cpp/src/base/impl/ServicesImpl.cpp
/*
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under the EUPL, Version 1.1 only (the "License");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * http://joinup.ec.europa.eu/software/page/eupl
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and
 * limitations under the Licence.
 */

#include "ServicesImpl.h"

#include "SocketException.h"
#include "ConnectionTimeout.h"
#include "JSON.h"
#include "RequestSocketImpl.h"
#include <iostream>
#include <sstream>
#include "../Requests.h"

// Using Visual Studio preprocessor.
// It must be improved in case of other compilers.
#ifdef _WIN32
	#include <process.h>
	#define GET_PROCESS_PID() _getpid()
#else
	#include <unistd.h>

	#define GET_PROCESS_PID() ::getpid()
#endif


using namespace std;

namespace cameo {

ServicesImpl::ServicesImpl() :
	m_context(1), m_timeout(0) {
}

ServicesImpl::~ServicesImpl() {
}

void ServicesImpl::setTimeout(int timeout) {
	m_timeout = timeout;
}

int ServicesImpl::getTimeout() const {
	return m_timeout;
}

long ServicesImpl::getPid() const {
	return GET_PROCESS_PID();
}

zmq::socket_t * ServicesImpl::createEventSubscriber(const std::string& endpoint, const std::string& cancelEndpoint) {

	zmq::socket_t * subscriber = new zmq::socket_t(m_context, ZMQ_SUB);

	vector<string> streamList;
	streamList.push_back(message::Event::STATUS);
	streamList.push_back(message::Event::RESULT);
	streamList.push_back(message::Event::PUBLISHER);
	streamList.push_back(message::Event::PORT);
	streamList.push_back(message::Event::KEYVALUE);
	streamList.push_back(message::Event::CANCEL);

	for (vector<string>::const_iterator s = streamList.begin(); s != streamList.end(); ++s) {
		subscriber->setsockopt(ZMQ_SUBSCRIBE, s->c_str(), s->length());
	}

	subscriber->connect(endpoint.c_str());
	subscriber->connect(cancelEndpoint.c_str());

	return subscriber;
}

zmq::socket_t * ServicesImpl::createOutputStreamSubscriber(const std::string& endpoint, const std::string& cancelEndpoint) {

	zmq::socket_t * subscriber = new zmq::socket_t(m_context, ZMQ_SUB);

	vector<string> streamList;
	streamList.push_back(message::Event::SYNCSTREAM);
	streamList.push_back(message::Event::STREAM);
	streamList.push_back(message::Event::ENDSTREAM);
	streamList.push_back(message::Event::CANCEL);

	for (vector<string>::const_iterator s = streamList.begin(); s != streamList.end(); ++s) {
		subscriber->setsockopt(ZMQ_SUBSCRIBE, s->c_str(), s->length());
	}

	subscriber->connect(endpoint.c_str());
	subscriber->connect(cancelEndpoint.c_str());

	return subscriber;
}

zmq::socket_t * ServicesImpl::createCancelPublisher(const std::string& endpoint) {

	zmq::socket_t * publisher = new zmq::socket_t(m_context, ZMQ_PUB);
	publisher->bind(endpoint.c_str());

	return publisher;
}

zmq::socket_t * ServicesImpl::createRequestSocket(const std::string& endpoint) {

	zmq::socket_t* socket = new zmq::socket_t(m_context, ZMQ_REQ);

	try {
		// Set the linger value to 0 to ensure that pending requests are destroyed in case of timeout.
		int value = 0;
		socket->setsockopt(ZMQ_LINGER, &value, sizeof(int));

		// Connect to the endpoint.
		socket->connect(endpoint.c_str());
	}
	catch (exception const & e) {
		throw SocketException(e.what());
	}

	return socket;
}

bool ServicesImpl::isAvailable(RequestSocketImpl * socket, int timeout) {

	string request = createSyncRequest();

	try {
		unique_ptr<zmq::message_t> reply = socket->request(request, timeout);

		if (reply.get() != nullptr) {
			return true;
		}
	}
	catch (const ConnectionTimeout&) {
		// The server is not accessible.
	}
	catch (...) {
		// Should not happen.
	}

	return false;
}

void ServicesImpl::sendSyncStream(RequestSocketImpl * socket, const std::string& name) {

	string request = createSyncStreamRequest(name);

	try {
		unique_ptr<zmq::message_t> reply = socket->request(request);
	}
	catch (const ConnectionTimeout&) {
		// The server is not accessible.
	}
	catch (...) {
		// Should not happen.
	}
}

void ServicesImpl::waitForStreamSubscriber(zmq::socket_t * subscriber, RequestSocketImpl * socket, const std::string& name) {

	// Poll subscriber.
	zmq_pollitem_t items[1];
	items[0].socket = static_cast<void *>(*subscriber);
	items[0].fd = 0;
	items[0].events = ZMQ_POLLIN;
	items[0].revents = 0;

	while (true) {
		sendSyncStream(socket, name);

		// Wait for 100ms.
		int rc = zmq::poll(items, 1, 100);
		if (rc != 0) {
			break;
		}
	}
}

void ServicesImpl::waitForSubscriber(zmq::socket_t * subscriber, RequestSocketImpl * socket) {

	// Poll subscriber.
	zmq_pollitem_t items[1];
	items[0].socket = static_cast<void *>(*subscriber);
	items[0].fd = 0;
	items[0].events = ZMQ_POLLIN;
	items[0].revents = 0;

	while (true) {
		isAvailable(socket, 100);

		// Wait for 100ms.
		int rc = zmq::poll(items, 1, 100);
		if (rc != 0) {
			break;
		}
	}
}

}
